### DISCUSSION TITLE
 "ASAP" deletion policy 

### DISCUSSION THREAD

**Discussion by rd4com** [modular/modular#565] [Ideas]
it might shows my ignorance:

in the case of being inside an hardware interrupt it might not be an usefull thing if someone want to control the timing of things like when dealing with sensors and microcontrollers,
specifically for some users in the case some operations need to be completed asap inside an interrupt before another one fire up.
Another use case might be when using ADCs or DACs , when the goal is to produce a signal as soon as the clock tick, in precise speed audio,motors, servos and robotics in general. 
might be great to be able to do something like @batch_deletion above a function or an expression

**Comment by drunkwcodes** [comment]
Do you have a concrete use case?

What programming language are you using for this feature?

**Reply by rd4com** [reply]
i might be wrong but i am quite confident that freedom to control it might be significant to some uses cases,
it might be of interest to ask to hardware/system users what they think about it,
because ai runs on hardware and sometimes data comes from sensors.
(i dont know if you would loses accuracy if time/clock-cycles were spent cleaning instead of executing in some cases)

https://docs.modular.com/mojo/programming-manual.html#behavior-of-destructors  
Even in an expression like a+b+c+d, 
Mojo destroys the intermediate expressions eagerly, 
as soon as they are no longer needed—it does not wait until the end of the statement.

**Comment by lattner** [comment]
Mojo gives you the ability to control this, by extending the lifetime.  You can do things like:

```
critical_section_start()  # Whatever this does
foo(a+b)
# the result of "a+b" is destroyed here.
critical_section_end()
```

or if you want the destruction outside the critical section, you can do:

```
critical_section_start()  # Whatever this does
let tmp = a+b
foo(tmp)
critical_section_end()

_ = tmp
# the result of "a+b" is destroyed here, after the last use of tmp.
```

**Comment by rd4com** [comment]
all right then fantastic if there is no issues i guess the topic is closed :+1:

**Comment by rd4com** [comment]
for completeness, yes it is in the documentation, i overlooked, apologizes:

https://docs.modular.com/mojo/programming-manual.html#lifetimes

You should not generally have to think about this, but if you have logic with inner pointers into members, you may need to keep them alive for some logic within the destructor or move initializer itself. You can do this by assigning to the _ “discard” pattern:

fn __del__(owned self):
    self.dump() # Self is still whole here

    consume(self.str1^)
    _ = self.str2
    # self.str2.__del__(): Mojo destroys str2 after its last use.

In this case, if consume() implicitly refers to some value in str2 somehow, this will ensure that str2 isn’t destroyed until the last use when it is accessed by the _ discard pattern.
